lib/IDE/Refactoring.cpp

```diff
@@ -5310,6 +5337,293 @@ class AsyncConverter : private SourceEntityWalker {
+      OS << "(";
+    }
+
+    llvm::interleave(
+        ReturnTypes, [&](Type Ty) { Ty->print(OS); }, [&]() { OS << ", "; });
+
+    if (ReturnTypes.size() > 1) {
+      OS << ")";
+    }
+
+    return ReturnType;
+  }
+
+  /// If the async alternative function is generic, return the type annotation
+  /// to the 'return' variable in the legacy function so that the generic
+  /// parameters of the legacy function are passed to the generic function.
+  /// Otherwise return the empty string.
+  /// For example for
+  /// \code
+  /// func foo<GenericParam>() async -> GenericParam {}
+  /// \endcode
+  /// we generate
+  /// \code
+  /// func foo<GenericParam>(completion: (T) -> Void) {
+  ///   detach {
+  ///     let result: GenericParam = await foo()
+  ///               <------------>
+  ///     completion(result)
+  ///   }
+  /// }
+  /// \endcode
+  /// The range returned by this funciton is annotated with \c <----->
+  std::string getResultTypeAnnotationIfNecessary() const {
+    if (FD->isGeneric()) {
+      auto ResultType = getAsyncFuncReturnType();
+      std::string ResultTypeAnnotation;
+      llvm::raw_string_ostream(ResultTypeAnnotation) << ": " << ResultType;
+      return ResultTypeAnnotation;
+    }
+    return "";
+  }
+
+public:
+  LegacyAlternativeBodyCreator(FuncDecl *FD, AsyncHandlerDesc HandlerDesc)
+      : FD(FD), HandlerDesc(HandlerDesc) {}
+
+  bool canRewriteLegacyBody() {
+    if (FD == nullptr || FD->getBody() == nullptr) {
+      return false;
+    }
+    if (FD->hasThrows()) {
+      assert(!HandlerDesc.isValid() && "We shouldn't have found a handler desc "
+                                       "if the original function throws");
+      return false;
+    }
+    switch (HandlerDesc.Type) {
+    case HandlerType::INVALID:
+      return false;
+    case HandlerType::PARAMS: {
+      if (HandlerDesc.HasError) {
+        // The non-error parameters must be optional so that we can set them to
+        // nil in the error case.
+        // The error parameter must be optional so we can set it to nil in the
+        // success case.
+        // Otherwise we can't synthesize the values to return for these
+        // parameters.
+        return llvm::all_of(HandlerDesc.params(),
+                            [](AnyFunctionType::Param Param) -> bool {
+                              return Param.getPlainType()->isOptional();
+                            });
+      } else {
+        return true;
+      }
+    }
+    case HandlerType::RESULT:
+      return true;
+    }
+  }
+
+  std::string create() {
+    assert(canRewriteLegacyBody() &&
+           "Cannot create a legacy body if the body can't be rewritten");
+    std::string Buffer;
+    llvm::raw_string_ostream OS(Buffer);
+
+    OS << "{\n"; // start function body
+    OS << "detach {\n";
+    switch (HandlerDesc.Type) {
+    case HandlerType::INVALID:
+      llvm_unreachable("Cannot be rewritten");
+      break;
+    case HandlerType::PARAMS: {
```

Comment:
These two cases are identical except for the `getCallToCompletionHandler` part... Could that just handle `Result` as well? ie. `HandlerDesc` is a `RESULT` return either `completion(.success(result))\n` or `"completion(.failure(error" << getCastToCustomErrorTypeIfNecessary() << "))\n"`

Reply:
Oh, yeah. Good idea. Definitely simplifies things.